data_model/src/query/mod.rs

```diff
@@ -250,303 +250,331 @@ mod model {
+macro_rules! impl_iter_queries {
+    ($ty:ty => $item:ty $(, $($rest:tt)*)?) => {
+        impl IterableQuery for $ty {
+            type Item = $item;
         }
 
-        impl From<$ty> for QueryOutputBox {
-            fn from(id: $ty) -> Self {
-                Self::Id(IdBox::$variant(id))
-            }
-        })+
+        $(
+            impl_iter_queries!($($rest)*);
+        )?
     };
+    // allow for a trailing comma
+    () => {}
 }
 
-macro_rules! from_and_try_from_value_identifiable {
-    ( $( $variant:ident( $ty:ty ), )+ $(,)? ) => { $(
-        impl TryFrom<QueryOutputBox> for $ty {
-            type Error = iroha_macro::error::ErrorTryFromEnum<QueryOutputBox, Self>;
-
-            fn try_from(value: QueryOutputBox) -> Result<Self, Self::Error> {
-                if let QueryOutputBox::Identifiable(IdentifiableBox::$variant(id)) = value {
-                    Ok(id)
-                } else {
-                    Err(Self::Error::default())
-                }
-            }
+/// Use a custom syntax to implement [`SingularQueries`] for applicable types
+macro_rules! impl_singular_queries {
+    ($ty:ty => $output:ty $(, $($rest:tt)*)?) => {
+        impl SingularQuery for $ty {
+            type Output = $output;
         }
 
-        impl From<$ty> for QueryOutputBox {
-            fn from(id: $ty) -> Self {
-                Self::Identifiable(IdentifiableBox::$variant(id))
-            }
-        } )+
+        $(
+            impl_singular_queries!($($rest)*);
+        )?
     };
+    // allow for a trailing comma
+    () => {}
 }
 
-from_and_try_from_value_idbox!(
-    PeerId(crate::peer::PeerId),
-    DomainId(crate::domain::DomainId),
-    AccountId(crate::account::AccountId),
-    AssetId(crate::asset::AssetId),
-    AssetDefinitionId(crate::asset::AssetDefinitionId),
-    TriggerId(crate::trigger::TriggerId),
-    RoleId(crate::role::RoleId),
-    // TODO: Should we wrap String with new type in order to convert like here?
-    //from_and_try_from_value_idbox!((DomainName(Name), ErrorValueTryFromDomainName),);
-);
-
-from_and_try_from_value_identifiable!(
-    NewDomain(crate::domain::NewDomain),
-    NewAccount(crate::account::NewAccount),
-    NewAssetDefinition(crate::asset::NewAssetDefinition),
-    NewRole(crate::role::NewRole),
-    Peer(crate::peer::Peer),
-    Domain(crate::domain::Domain),
-    Account(crate::account::Account),
-    AssetDefinition(crate::asset::AssetDefinition),
-    Asset(crate::asset::Asset),
-    Trigger(crate::trigger::Trigger),
-    Role(crate::role::Role),
-);
-
-impl<V: Into<QueryOutputBox>> From<Vec<V>> for QueryOutputBox {
-    fn from(values: Vec<V>) -> Self {
-        Self::Vec(values.into_iter().map(Into::into).collect())
-    }
+impl_iter_queries! {
+    FindRoles => crate::role::Role,
+    FindRoleIds => crate::role::RoleId,
+    FindRolesByAccountId => crate::role::RoleId,
+    FindPermissionsByAccountId => crate::permission::Permission,
```

Comment:
This information should be provided in the schema, so that SDK can auto-generate these mappings too.

For a different PR, I guess.

Reply:
Opened #4902

Reply:
should be done sooner rather than later